### ISSUE TITLE
Missing columns "Heat" and "Priority"

### ISSUE THREAD

**Issue by ThiloPetry** [drachels/moodle-mod_hotquestion#46]
Hello,

After our update to 3.9.3. (Build: 2020060603) the two columns are missing in the activity.
We haven't changed anything in the settings:
![image](https://user-images.githubusercontent.com/33430236/104729612-fa0eef80-5738-11eb-93a1-ce42bafed832.png)
But the result now looks like this:
![image](https://user-images.githubusercontent.com/33430236/104729730-26c30700-5739-11eb-90e8-a4d371ee3789.png)
We do not know why the columns disappeared. 
On 3.8 everything worked fine.

**Comment by drachels** [comment]
Sorry, there is not a Hot Question version 3.9.3 yet. Do you mean version 3.9.0+ (Build: 2020060603)?
I would suggest that you go ahead and upgrade to the latest Hot Question v3.9.1 (Build: 2020121700) and see if it fixes the problem.

**Comment by ThiloPetry** [comment]
Oh sorry, 
My mistake. 
We are now on plugin version 3.9.1 (Build: 2020121700) 
The columns are still missing.
It is also interesting that when I set the visibility to "Nein" and save, this step is not saved. 
The visibility is always "Ja" without the columns being displayed. 
We have as Moodle version 3.9.3 and use PostgreSQL 9.6.20

**Comment by drachels** [comment]
It will take a while before I can try to duplicate this. My only server that has PostgreSQL 9.6.20, currently only has PHP 7.0 on it. I've been putting off doing an upgrade, but have started to do so now.

**Comment by drachels** [comment]
Sorry to take so long, but there appears to be something wrong with your installation of Hot Question, because one of your settings, Allow authors name to be hidden, appears to be missing.
![PostgreSQLHQEdit1](https://user-images.githubusercontent.com/1018249/105537720-f4288980-5cb7-11eb-93cf-ba9f9b010785.png)
If you have something to look at your moodledatabase with, such as adminer, verify that you have the correct tables in mdl_hotquestion.
![PostgreSQLHQtable1](https://user-images.githubusercontent.com/1018249/105538390-f0e1cd80-5cb8-11eb-8127-991adb81542d.png)

**Comment by ThiloPetry** [comment]
We are missing the two columns teacherpriorityvisibility and heatvisibility.
This probably explains where the error comes from. 
Any tips on how we can add the columns?

**Comment by drachels** [comment]
You should be able to manually add the two missing fields using phpMyAdmin or the Moodle plugin adminer. Be sure and make a backup copy of the table first, just in case.
It would be even easier, if you do not have too many Hot Question activities, to just delete the plugin, then reinstall the latest version.
As an alternative if you have phpMyAdmin or adminer installed but do not want to try and fix it yourself, you can send me the link to your Moodle via email, and give me a temporary admin account, and I will try to fix it for you. my email is drachels@drachels.com

**Comment by ksteitz** [comment]
Thanks for you support, head of Thilos Team speaking here. We have added the two columns, but as expected it doesnt help for existing activities, because they have only "null" in the columns. Creating a new activity now shows "Allow authors name to be hidden" and columns "Heat" and "Priority" are shown. 
Reinstalling isn´t an option for us, thus we have 1536 hotquestion-activities in our system. 
There has to be either a for us on PostgreSQL-DB working upgrade-script from Plugin-Version 3.8 to 3.9 or kind of SQL-Statement for filling the two manually created tables with appropriate data from the existing activities. 
As long as we do not have that, we have to stay using the Version 3.8, which is working fine for us. 

Are there any other systems using PostgreSQL on current Plugin-Version 3.9, which were able to update from an earlier version (like 3.8) without problems?

kind regards
Klaus

**Comment by drachels** [comment]
So far, testing upgrades from last 3.8.2 to latest 3.9.1, I don't have a problem and I am getting the correct results. Could you tell me specifically which 3.8.x version you are trying to upgrade from? (Three is a 3.8.0, 3.8.1, and 3.8.2.) Then also tell me which of the 3.9.x's you are upgrading to so that I can try to specifically duplicate your environment for testing. Should be try to go to 3.9.1.

**Comment by ksteitz** [comment]
We ugraded from Plugin-Version 3.8.0 (Build: 2019112100) to 3.9.0+ (Build: 2020060603) (on Moodle 3.9.3 with PostgreSQL 9.6.20). Later from 3.9.0+ to 3.9.1. 
I also tried just on a new copy of our production system upgrading directly from 3.8.0 (Build: 2019112100) to 3.9.1 (Build: 2020121700). The columns are still missing onexisting and new activities. 
Also the strings are not identified correctly and are only in brackets [[]], e.g. [[questionlabel]]. None of the "new strings" (since 3.8) are recognized. Although i translated them severals months ago in AMOS to german.

**Comment by drachels** [comment]
I'm not sure why the translation strings are not working, but I will watch for that while working on the columns. I have a couple of ideas about the columns though, and will see what I can do about that. I will have to go back and reset my PostgreSQL server version of Moodle 3.9.x first.

**Comment by drachels** [comment]
Sorry to take so long with this, but I just have not had much spare time for Moodle the past two weeks.
I went all the way back to HotQuestion v3.3.2 to try and track down were your problems might be coming from.

-  Starting with version 3.3.2, which does not have teacher priority column but does have Heat, Remove, and Approved for teachers and admin. Only Heat for students.
- HQ 3.4.0 added Priority column with data stored in mdl_hotquestion_questions table.
- HQ 3.4.1. No new columns. Works as expected. Nothing out of place.
- HQ 3.5.0. Some new GDPR strings.
- HQ 3.5.1. GDPR released.
- HQ3.5.2. Lots of stuff added. Priority added to download. PostgreSQL fix for quotes, concat and timestamp.
- HQ3.5.2+. Nothing out of place.
- HQ3.6.0. Nothing out of place.
- HQ3.7.0. Nothing out of place.
- HQ3.8.0. Added setting for Priority and Heat visibility. Defaults are both yes, show the column. New setting is in the mdl_hotquestions table.
- HQ3.8.1. Nothing out of place.
- HQ3.8.2. Added settings to allow changing all the column header label names. questionlabel, teacherprioritylabel, heatlabel, and qpprovallabel.
- HQ3.9.0.00. Added a heatlimit count. Had to go manually set existing HQ's to have a number of votes as they are set to 0 by default.
- HQ3.9.0.01+. Added capability to remove a vote and use it elsewhere.
- HQ3.9.0.02. Fixed remove when there are no votes left.
- HQ3.9.0.03+. Everything works as expected. Heat votes are working and both columns are visible to students.
- HQ3.9.1. Added setting to allow authors name to be hidden from other students. Admin/teachers still see it.

I did the first five upgrades in English(en-us) and the remainder using German(de) and did NOT see any string errors. After each upgrade everyone, students, teachers and admins, all could see the columns they were supposed to see.

**Comment by drachels** [comment]
Since I did not find any problems going directly from one version to another, I will next check to see if leaving out the first three v3.9.0, builds 2020060600, 2020060601, and 2020060602, are the cause of your problems.

**Comment by drachels** [comment]
I just am not able to do any upgrade combination that does not correctly make the database tables, nor can I make it leave out strings such that you would get the string only in brackets [[]], e.g. [[questionlabel]]

I do get a situation where an upgrade from HQ3.8.0 to any higher versions does make the Heat column become hidden in existing HotQuestion activities. This is due to the file located in, yourmoodle/mod/hotquestion/db/upgrade.php which currently defines the heatlimit to be zero. Changing this to a positive integer, e.g. 5, will make existing HQ activity heat columns to be automatically visible, without a teacher or admin having to edit the activity setting.

**Comment by ksteitz** [comment]
Good news, we have now solved the problem! 
It seems to have been caused by the fact that we had installed a development version of the plugin as we were retrieving a specific commit from the GIT repository: 3.8.0 (Build: 2019112100; de9075a2921bfd0cc820220e24773271d08eefe5). When first implemented, this version wasn't a dev-version but only became one later. However, in the upgrade.php for the newer version, this exact development version was disregarded.
The check is for “oldversion < 2019112100” and we were exactly on version 2019112100, so we weren't covered in the upgrade routine.
With a few little hacks we could now fix this and then upgrade to the currently newest version 2023040701, and also get the columns heat and priority back. So this issue can be marked as closed.
But we discovered a small new issue, that only shows up using PostgreSQL (and not with mySQL), i open a new issue for that. 

Thanks and cheers!
Klaus
